src/markdownEngine.ts

```diff
@@ -2,70 +2,71 @@
 
 import { extensions, workspace, WorkspaceConfiguration } from 'vscode';
 import { slugify } from './util';
+import { MarkdownContribution, MarkdownContributionProvider, getMarkdownContributionProvider } from './markdownExtensions'
 
 class MarkdownEngine {
-    public md?;
+    public cacheMd;
 
+    public async getEngine() {
+        if (!this.cacheMd) {
+            this.cacheMd = await this.newEngine();
+        }
+        return this.cacheMd;
+    }
+
+    public contributionsProvider = getMarkdownContributionProvider();
     private _slugCount = new Map<string, number>();
 
     constructor() {
-        this.initMdIt();
+        this.contributionsProvider.onContributionsChanged(() => {
+            this.cacheMd = null;
+            this.getEngine();
```

Comment:
Why not `this.cacheMd = this.newEngine();`

Reply:
Both `initMdIt` (old version) and `newEngine` (new version) are async function, while constructor can't be async. `util/mdHeadingToPlaintext` will use the cached engine, I just keep the original behavior: a engine will be created asynchronously in constructor. In fact, I don't have a clear understand about `mdHeadingToPlaintext`.

Reply:
I mean if you expand `getEngine` here, you get exactly `this.cacheMd = this.newEngine()` (without `await`) as `this.cacheMd == null` 😳.

`mdHeadingToPlaintext` is a naive implementation of "Markdown -> HTML -> text", which is a part of GitHub/GitLab-style `slugify` function.

Reply:
Well, this should be a bug, but `initMdIt` (also a async function, in old version) is called in this way too.